### ISSUE TITLE
CurrentRole  PR: Scrap Some Boilerplate

### ISSUE THREAD

**Issue by Montmorency** [digitallyinduced/ihp#703]
An idea I've had for a while is to rationalize some of the code surrounding the LoginSupport. For instance at the moment there is a bit of repetition in `IHP.LoginSupport.Helper.Controller` for `User` and `Admin` types. For instance the repetition includes `currentUser`/`currentAdmin` and the logic for handling defining sessions. The PR I am working on would propose this: we introduce a `CurrentRole` type family in LoginSupport. 

```
type family CurrentRole a
```

The if the user has records for different application User types (User, Admin, Provider etc.) we create instances in the Application/Controller Helper e.g.:

```
type instance CurrentRole User = User
type instance CurrentRole Admin = Admin
type instance CurrentRole Provider = Provider
```

then the type signatures of the helper functions are amended to make them completely generic: 
```
currentRole :: forall roletype.(?context :: ControllerContext, HasNewSessionUrl (CurrentRole roletype), Typeable (CurrentRole roletype), roletype ~ (CurrentRole roletype)) => roletype
  currentRole = fromMaybe (redirectToLogin (newSessionUrl (Proxy @roletype))) currentInstallerOrNothing

currentRoleOrNothing :: forall roletype.(?context :: ControllerContext, Typeable CurrentRole roletype, roletype ~ CurrentRole roletype) => (Maybe roletype)
currentRoleOrNothing = case unsafePerformIO (maybeFromContext @(Maybe CurrentRole roletype)) of
                                      Just a -> a
                                      Nothing -> error "currentRoleOrNothing: initAuthentication @Role has not been called in initContext inside FrontController of this application."
```

They just have to fetch the CurrentRole from the context, unwrap the data structure and query the Record. 

As I visualize it this PR would not break backwards compatibility, because the old `CurrentUserRecord` and related helper functions would still be available it just requires adding the generic `CurrentRole` typed functions and appropriate type family instances. 

Any comments suggestions much appreciated.

**Comment by mpscholten** [comment]
In general I'm happy with refactoring this. The reason we have the `CurrentUserRecord` type instances is mainly to avoid issues with type inference, e.g. without the type family ghc would not be able to know what specific type is needed when calling `currentUser`. Then all call-sites to `currentUser` would need to have type annotations, like `currentUser @User`. Doesn't the `currentRole` helper have the same type inference issue?

**Comment by Montmorency** [comment]
I should maybe be a little clearer on the intended use cases. I have thought about simply having an single global user type in the application with an ENUM record field that specifies that "user's" role (or privilege etc.) but I think there are many cases where the types of user are sufficiently different that separate tables are required for each of the user types this is where Roles would come in.  

I'm trying to design the "Roles" to try and make it very simple to write:

         1) User Type specific rendering In Views
         2) DB access on different User Types (e.g. `provider <- fetch currentRoleID @Provider`)
         3) Generic Controller `ensureIsRole @Admin` functions.  

(w.r.t to type annotations I think these will still be necessary in some cases e.g. on database access and initializing sessions. For instance at the moment I am handling the creation of sessions for different user Types like this:

```
instance Controller SessionsController where
    action NewSessionAction = Sessions.newSessionAction @User
    action CreateSessionAction = Sessions.createSessionAction @User
    action DeleteSessionAction = Sessions.deleteSessionAction @User

    action NewProviderSessionAction = Sessions.newSessionAction @Provider
    action CreateProviderSessionAction = Sessions.createSessionAction @Provider
    action DeleteProviderSessionAction = Sessions.deleteSessionAction @Provider
```

As an example let's say we Have a `User` Type and a `Provider` Type and they both can access the same view of a `Store`. I would like to have a single `Store` endpoint for the view and customize that single view based on what user type (if any) is logged in. The motivation is the `Store` shares most of the same attributes but I would like different slightly different data presented to the `User` and the `Provider`. I also think for site maintenance this is an attractive design because all rendering logic with respects to stores is contained in a single StoreView.hs. Also the way functions are annotated it can be made explicit what will be rendered to the different user types.

Without Roles at the moment I can do a couple things. Either I split the views for each user type. This has a few drawbacks: it requires two endpoints for each user type with duplication of HTML and more overhead.

 I can also write `renderRoleSpecific` HTML helper functions in the view with type signatures like this:

```
= [hsx|
            renderRoleSpecific (getCurrentUserOrNothing) (getCurrentProviderOrNothing)
   |]
...

renderRoleSpecific :: Maybe User -> Maybe Provider -> Html
renderRoleSpecific (Just User) Nothing = [hsx|user specific|]
renderRoleSpecific Nothing (Just Provider) = [hsx|provider specific|]
renderRoleSpecific Nothing Nothing = [hsx||]
```

Having written thought about this some more I'm starting to think a GADT might be better than a type family. Something like:

```
data Role a where
    UserRole :: Role User
    ProviderRole :: Role Provider

renderRole :: Role a -> Html
renderRole UserRole _ = [hsx||]
renderRole ProviderRole _ = [hsx||]
 ```

A similar pattern might work for the `ensureIsRole AdminRole` etc..


The only "dirty" part of the whole process I think comes with reading the session key and mapping the session string to the user type which might require an `unsafeCoerce`.

I'll keep playing with this but if it sounds like a sensible design pattern let me know. (Also if I'm missing a much simpler way of handling multiple user types!)

**Comment by mpscholten** [comment]
Is it possible to have e.g. a single table for storing user and provider? Then this slack thread here could be useful: https://ihpframework.slack.com/archives/C01DQE0F4F8/p1610390775053700

**Comment by Montmorency** [comment]
Yes, that discussion has a number of similarities! Both this thread and that discussion turn on a choice between two designs: 

 1) Having a single table and differentiating the "Types" within that table (User Types or Weapon types) via an ENUM attribute. 

 2) Splitting the types between different tables. 

The first option would work in many cases but as discussed if the attributes between the different Types within a table are divergent you end up populating the record fields with lots of `Nothing`s. It could also lead to problems ensuring the certain types of the ENUM don't have fields populated for properties they should never possess. 

The other difference is the authentication here needs to tie in with the IHP sessions logic more closely and that the notion of a generic "Roles" would  help authorization/authentication, and views across a wider range of applications. Many applications will require  A User, Admin, Supplier, Department etc. Role so the more generic (without getting too complicated) authorization is it can save a lot of boilerplate (e.g. customizing views example I gave earlier). 

This post gives a really nice discussion of another "Role" based authentication problem in a real setting [Authorization](https://ocharles.org.uk/posts/2019-08-09-who-authorized-these-ghosts.html).
